libtenzir/src/table_slice.cpp

```diff
@@ -93,33 +95,245 @@ create_table_slice(const std::shared_ptr<arrow::RecordBatch>& record_batch,
+  auto builder = series_builder{};
+  for (auto i = int64_t{0}; i < map->length(); ++i) {
+    const auto start = map->value_offset(i);
+    const auto end = start + map->value_length(i);
+    auto rec = builder.record();
+    for (auto j = start; j < end; ++j) {
+      const auto key = keys_string->Value(j);
+      const auto value = values_string->Value(j);
+      rec.field(key, value);
+    }
+  }
+  auto s = builder.finish_assert_one_array();
+  auto r = std::dynamic_pointer_cast<arrow::StructArray>(s.array);
+  TENZIR_ASSERT(r);
+  return r;
+};
+
+/// Tries to upgrade some column types to types supported by Tenzir, by
+/// recursively checking their types and casting.
+auto upgrade_arrays(const std::shared_ptr<arrow::Array>& array)
+  -> std::shared_ptr<arrow::Array> {
+  array->type();
+  switch (array->type_id()) {
+    case arrow::Type::MAP: {
+      const auto map_array = std::static_pointer_cast<arrow::MapArray>(array);
+      auto as_struct = try_convert_map_to_struct_array(map_array);
+      if (as_struct) {
+        return as_struct;
+      }
+      return array;
+    }
+    case arrow::Type::STRUCT: {
+      auto struct_array = std::static_pointer_cast<arrow::StructArray>(array);
+      auto new_arrays = std::vector<std::shared_ptr<arrow::Array>>{};
+      auto new_fields = std::vector<std::shared_ptr<arrow::Field>>{};
+      new_arrays.reserve(struct_array->num_fields());
+      for (int i = 0; i < struct_array->num_fields(); ++i) {
+        const auto& array = struct_array->field(i);
+        const auto& field = struct_array->struct_type()->field(i);
+        auto new_array = upgrade_arrays(struct_array->field(i));
+        auto new_field = field;
+        if (new_array != array) {
+          new_field
+            = arrow::field(field->name(), new_array->type(), field->nullable());
         }
-      },
-      [&](const arrow::ListArray& la) noexcept {
-        check_col(check_col, *la.values());
-      },
-      [&](const arrow::MapArray& ma) noexcept {
-        check_col(check_col, *ma.keys());
-        check_col(check_col, *ma.items());
-      },
-      [](const arrow::Array&) noexcept {},
-    };
-    match(column, f);
-  };
-  for (const auto& column : record_batch.columns()) {
-    check_col(check_col, *column);
+        new_arrays.push_back(std::move(new_array));
+        new_fields.push_back(std::move(new_field));
+      }
+      return make_struct_array(struct_array->length(),
+                               struct_array->null_bitmap(),
+                               std::move(new_fields), std::move(new_arrays));
+    }
+    case arrow::Type::LIST: {
+      auto list_array = std::static_pointer_cast<arrow::ListArray>(array);
+      auto values = upgrade_arrays(list_array->values());
+      return check(arrow::ListArray::FromArrays(
+        *list_array->offsets(), *values, arrow::default_memory_pool(),
+        list_array->null_bitmap(), list_array->null_count()));
```

Comment:
```suggestion
        *list_array->offsets(), *values, arrow::default_memory_pool(),
        list_array->null_bitmap(), list_array->data().null_count));
```
You always want to go through `data()` for the null count.

By the way, I think this code doesn't handle offsets correctly, but that should be fine when reading in Parquet.